src/udev/net/ethtool-util.c

```diff
@@ -473,15 +473,16 @@ static int set_slinksettings(int *fd, struct ifreq *ifr, const struct ethtool_li
         struct {
                 struct ethtool_link_settings req;
                 __u32 link_mode_data[3 * ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32];
-        } ecmd = {
-                .req.cmd = ETHTOOL_SLINKSETTINGS,
-        };
+        } ecmd;
         unsigned int offset;
         int r;
 
         if (u->base.cmd != ETHTOOL_GLINKSETTINGS || u->base.link_mode_masks_nwords <= 0)
                 return -EINVAL;
 
+        memset(&ecmd, sizeof(ecmd), 0);
+        memcpy(&ecmd.req, &u->base, sizeof(ecmd.req));
```

Comment:
Hmm, the u->base and ecmd.req should have the same type, no? If so, please just use normal assignment, not memcpy(). i.e.:

```c
ecmd.req = u->base;
````

Reply:
I get your first change to eliminate the memset. Regarding the memcpy in 484, ecmd.req and u->base  are _struct ethtool_link_settings_, not pointers to one. Does C now allow wholesale assignment of struct contents?

Looking at the comparable function from ethtool.c, they're doing a memcpy also, and not being abreast of the latest changes in C, I felt I was safe:
```
static int
do_ioctl_slinksettings(struct cmd_context *ctx,
                       const struct ethtool_link_usettings *link_usettings)
{
        struct {
                struct ethtool_link_settings req;
                __u32 link_mode_data[3 * ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32];
        } ecmd;
        unsigned int u32_offs;

        /* refuse to send ETHTOOL_SLINKSETTINGS ioctl if
         * link_usettings was retrieved with ETHTOOL_GSET
         */
        if (link_usettings->base.cmd != ETHTOOL_GLINKSETTINGS)
                return -1;

        /* refuse to send ETHTOOL_SLINKSETTINGS ioctl if deprecated fields
         * were set
         */
        if (link_usettings->deprecated.transceiver)
                return -1;

        if (link_usettings->base.link_mode_masks_nwords <= 0)
                return -1;

        memcpy(&ecmd.req, &link_usettings->base, sizeof(ecmd.req));
        ecmd.req.cmd = ETHTOOL_SLINKSETTINGS;
    . . .
```
I'm going to try pushing the first change. Please bear with me, as this is my first experience with Github and pull requests.

Reply:
yes, C has supported assigning full struct since about always. memcpy() is not needed

Reply:
Indeed, I was sure this doesn't work. Glad to see that it does. So both set_glinksettings and get_glinksettings can be adjusted.

Reply:
comparison of structs is not allowed in C, because the padding in structs makes this useless in the general case. But assignment of structs has always been fine

Reply:
Note that I included a fix-up for this in #7816